packages/teuchos/parameterlist/src/Teuchos_VerboseObjectParameterListHelpers.cpp

```diff
@@ -117,19 +117,81 @@ void Teuchos::readVerboseObjectSublist(
     voSublist,VerbosityLevel_name,VerbosityLevel_default
     );
+  // the default file string is nothing
   if (outputFileStr==OutputFile_default) {
     *oStream = null;
   }
+  // if a file is specified then output to an fstream
   else {
-    RCP<std::ofstream>
-      oFileStream = rcp(new std::ofstream(outputFileStr.c_str()));
-    TEUCHOS_TEST_FOR_EXCEPTION_PURE_MSG(
-      oFileStream->eof(), Exceptions::InvalidParameterValue,
-      "Error, the file \"" << outputFileStr << "\n given by the parameter\n"
-      "\'" << OutputFile_name << "\' in the sublist\n"
-      "\'" << voSublist.name() << "\' count not be opened for output!"
-      );
-    *oStream = fancyOStream(rcp_implicit_cast<std::ostream>(oFileStream));
+
+    // JJE: 14 March 2019
+    //  A fix for file output of an VerboseObject.
+    //
+    // This step is very important. With filestreams it does not make
+    // sense for multiple MPI ranks to open the same file.  Nor,
+    // does it seem inline with the OStream model that each stream
+    // represent a unique file. Perhaps, if that functionality is desired
+    // then the file name could be suffixed with the MPI Rank.
+    //
+    // A fundamental flaw with this implementation is that we have no knowledge
+    // of a communicator on which this OStream belongs. That makes the idea
+    // of using a rank ambiguous.
+    //
+    // The code below was added, and uses COMM_WORLD, because as-is
+    // this functionality was fundamentally broken. Without restricting
+    // the stream to a single rank, two severe consquences follow:
+    //   1) Each MPI process opens the file, which is not scalable;
+    //   2) Moreover, each MPI process *writes* to the file. Which
+    //      can give the illusion that things are working, if each
+    //      process writes exactly the same information (e.g., solver
+    //      convergence information for a bulk synchronous solve).
+    //      This introduces a terrible scalability problem, as the
+    //      filesystem is then tasked with coping with concurrent writes
+    //      to the same shared file, which is should make you cry a little.
+    //
+    // The resolution, is two fold:
+    //   1st, construct the ostream as a regular wrapper around cout
+    //   2nd, restrict the file creation and opening to a single process
+    //   Finally, map all ostreams except the fstream one to
+    //    a blackhole. Ensuring each rank has a functional stream
+    //    but that only one actually emits data to disk
+    //
+
+    // this could be a BlackHole, but calling setOutputToRootOnly does slightly
+    // more than simply blackhole output, it also disabled buffering across processes
+    *oStream = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
+
+    // Until we resolve OS streams that are communicator aware, use rank 0
+    const int outputFileMPIRank = 0;
+
+    #if defined(HAVE_TEUCHOS_MPI)
+    const int rank = Teuchos::GlobalMPISession::getRank();
```

Comment:
@jjellio What happens if process 0 in MPI_COMM_WORLD is no longer part of the current MPI communicator?  No output?

Reply:
@jhu2 This is a broader problem.  OStreams (and VerboseObject) are communicator-dumb. We discussed this above, the fix for that is more general and invasive.  Effectively, we need OStreams to become linked to a communicator.

Currently, when you blackhole a communicator (setOutputToRootOnly) this is actually using MPI_COMM_WORLD (from Teuchos GlobalMPISession ).  So existing output stuff is just as vulnerable to this as the fstream business. (Stratimikos is setting output to root only FYI).

I think this headache plagues MueLu more than anyone else.

As-is, this seems to work, since most (all?) codes like to use `setOutputToRootOnly` and guru like the MueLu devs have worked hard to make sure Rank 0 doesn't disappear.

Reply:
@jjellio and @jhux2,

I actually think this could be pretty easy to fix.  You just add a virtual function to this interface with a default implementation:

```
  virtual bool getVerbObjCommRank() const { return Teuchos::GlobalMPISession::getRank(); }
```

then subclasses that know they are parallel objects and have a `Teuchos::Comm<int>`  object (call it `comm_`) can define this as:

```
  virtual bool getVerbObjCommRank() const override { return comm_->getRank(); }
```

What do you think?

NOTE: We don't want to overload virtual functions, hence the longer name `getVerbObjCommRank()`.

Reply:
Clever!

Yes, that would default to what I have now.

I am not sure what Stratimikos/ThyraLinearOp expose in terms of communicators.

Reply:
@jjellio said:

> I am not sure what Stratimikos/ThyraLinearOp expose in terms of communicators.

That is true.  Since Thyra is an ANA (abstract numerical algorithm) interface, the objects don't return comms because that allows the client to assume a runtime model for the machine.  But given the practicality of comms in cases like this, we might consider adding a virtual function for returning an `RCP<const Teuchos::Comm<int>>` object with a default `null` implementation.

But that can happen later. For now, hopefully this PR will make things a little better.

But again, note that the fallback that gives complete control by the client is to create your own `Teuchos::FancyOStream` object wrapping any `std::ostream` subclass you want and then you can make it do anything you want.  It is just this ParameterList construction case that is the challenge.